superset/db_engine_specs/base.py

```diff
@@ -456,9 +424,15 @@ def apply_limit_to_sql(cls, sql: str, limit: int, database: "Database") -> str:
             )
             return database.compile_sqla_query(qry)
 
-        if cls.limit_method == LimitMethod.FORCE_LIMIT:
-            parsed_query = sql_parse.ParsedQuery(sql)
-            sql = parsed_query.set_or_update_query_limit(limit)
+        if LimitMethod.FORCE_LIMIT:
+            engine = cls.get_engine(database)
+            url_type = str(engine.url).split(':')[0]
+            parsed_query = sql_parse.ParsedQuery(sql, uri_type = url_type)
+            print(dir(engine.url))
+            if url_type in ['teradatasql','teradata']:
+                sql = parsed_query.set_or_update_query_limit_td(limit)
```

Comment:
My main concern with this approach is this. We shouldn't have Teradata-specific logic living inside this file, since all database specific logic should live in the custom DB engine spec (`superset/db_engine_specs/teradata.py` in this case).

Reply:
We have tried to do changes in teradata.py by moving the code from sql_parse file to teradata.py apply_limit_to_sql In base.py file to fetch set_or_update_query_limit_td in base.py file
by modifying the apply_limit_to_sql with exception.

This is not working as teradata file is referring the base file already hence both files are referring each other causing the issue.
[teradata_modified.txt](https://github.com/apache/superset/files/6500734/teradata_modified.txt)


It seems like SQL Lab is hit first which refer base file which finally refer sql_parse file, please let me know if this understanding is correct.


Please find the attached code for teradata file and let us know what changes can be done in files to successfully execute the same.

Reply:
We have done minimal changes in code for SQL_Parse and Base File only creating functions (set_or_update_query_limit_td, extract_limit_from_query_td) in SQL_Parse and slight modification mentioned below in apply_limit_to_sql to call set_or_update_query_limit_td.

We have also tested the same and it looks good by limiting the rows in SQL LAB.

This changes are required as new function if passed in Teradata file cannot be called in Base due to files calling each other.